substrate/frame/contracts/src/wasm/runtime.rs

```diff
@@ -2605,27 +2641,175 @@ pub mod env {
 		call_ptr: u32,
 		call_len: u32,
 	) -> Result<ReturnCode, TrapReason> {
-		use frame_support::dispatch::{extract_actual_weight, GetDispatchInfo};
+		use frame_support::dispatch::GetDispatchInfo;
 		ctx.charge_gas(RuntimeCosts::CopyFromContract(call_len))?;
 		let call: <E::T as Config>::RuntimeCall =
 			ctx.read_sandbox_memory_as_unbounded(memory, call_ptr, call_len)?;
-		let dispatch_info = call.get_dispatch_info();
-		let charged = ctx.charge_gas(RuntimeCosts::CallRuntime(dispatch_info.weight))?;
-		let result = ctx.ext.call_runtime(call);
-		let actual_weight = extract_actual_weight(&result, &dispatch_info);
-		ctx.adjust_gas(charged, RuntimeCosts::CallRuntime(actual_weight));
-		match result {
+		ctx.call_dispatchable(call.get_dispatch_info(), |ext| ext.call_runtime(call))
+	}
+
+	/// Execute an XCM message locally, using the contract's address as the origin.
+	/// This is equivalent to dispatching `pallet_xcm::execute` through call_runtime, except that
+	/// the function is called directly instead of being dispatched.
+	///
+	/// # Parameters
+	///
+	/// - `msg_ptr`: the pointer into the linear memory where the XCM message is placed.
+	/// - `msg_len`: the length of the XCM message in bytes.
+	///
+	/// # Return Value
+	///
+	/// Returns `ReturnCode::Success` when the XCM message was successfully executed. When the XCM
+	/// execution fails, `ReturnCode::CallRuntimeFailed` is returned.
+	#[unstable]
+	fn xcm_execute(
+		ctx: _,
+		memory: _,
+		msg_ptr: u32,
+		msg_len: u32,
+	) -> Result<ReturnCode, TrapReason> {
+		use crate::xcm::{CallOf, WeightInfo, XCM};
+		use frame_support::dispatch::DispatchInfo;
+		use xcm::VersionedXcm;
+
+		ctx.charge_gas(RuntimeCosts::CopyFromContract(msg_len))?;
+		let message: VersionedXcm<CallOf<E::T>> =
+			ctx.read_sandbox_memory_as_unbounded(memory, msg_ptr, msg_len)?;
+
+		let max_weight = ctx.ext.gas_meter().gas_left();
+		let max_weight =
+			max_weight.max(<<E::T as Config>::Xcm as XCM<E::T>>::WeightInfo::execute());
+		let dispatch_info = DispatchInfo { weight: max_weight, ..Default::default() };
```

Comment:
I am computing the `max_weight` argument here, used by xcm_execute, I think it's easier to compute it internally here rather than having to specify it in the contract